django/core/validators.py

```diff
@@ -413,15 +413,15 @@ def __init__(self, max_digits, decimal_places):
 
     def __call__(self, value):
         digit_tuple, exponent = value.as_tuple()[1:]
-        decimals = abs(exponent)
-        # digit_tuple doesn't include any leading zeros.
-        digits = len(digit_tuple)
-        if decimals > digits:
-            # We have leading zeros up to or past the decimal point. Count
-            # everything past the decimal point as a digit. We do not count
-            # 0 before the decimal point as a digit since that would mean
-            # we would not allow max_digits = decimal_places.
-            digits = decimals
+        if exponent >= 0:
+            digits = len(digit_tuple) + exponent
```

Comment:
Isn't this off by one? If I construct

```
x = Decimal("1.1e1")
val = DecimalValidator(2, 20)
```

then `val(x)` throws, even though `int(x) == 11`

Reply:
No but there is a different issue. Your example fails in the same way on both master and on this branch with:

`AssertionError: Validation of '11' failed. Error message was: ['Ensure that there are no more than -18 digits before the decimal point.']`

The problem is that it doesn't make any sense for `max_digits` to be > `decimal_places`. Not sure what the correct fix is for that.

Reply:
You can see the system check for it [here](https://github.com/django/django/blob/ff17ef1adab864d2c165d50501fa0b0684808d07/django/db/models/fields/__init__.py#L1510). I think we should add it to the docs.

Reply:
You're right, I misunderstood what was going on here. For some reason (I think because I hadn't loaded settings) my exceptions weren't printing messages, which didn't help.

Reply:
Maybe some comments would help to improve readability of these calculations?